src/python/pants/goal/auxillary_goal.py

```diff
@@ -0,0 +1,58 @@
+# Copyright 2024 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from typing import ClassVar
+
+from pants.base.exiter import ExitCode
+from pants.base.specs import Specs
+from pants.build_graph.build_configuration import BuildConfiguration
+from pants.engine.goal import GoalSubsystem
+from pants.engine.unions import UnionMembership
+from pants.init.engine_initializer import GraphSession
+from pants.option.options import Options
+from pants.option.scope import ScopeInfo
+
+
+@dataclass
+class AuxillaryGoalContext:
+    """Context passed to a `AuxillaryGoal.run` implementation."""
+
+    build_config: BuildConfiguration
+    graph_session: GraphSession
+    options: Options
+    specs: Specs
+    union_membership: UnionMembership
+
+
+class AuxillaryGoal(ABC, GoalSubsystem):
+    """Configure a "auxillary" goal which allows rules to "take over" Pants client execution in lieu
+    of executing an ordnary goal.
+
+    Only a single auxillary goal is executed per run, any remaining goals/arguments are passed
+    unaltered to the auxillary goal. Auxillary goals have precedence over regular goals.
+
+    When multiple auxillary goals are presented, the first auxillary goal will be used unless there is a
+    auxillary goal that begin with a hyphen (`-`), in which case the last such "option goal" will be
+    prioritized. This is to support things like `./pants some-builtin-goal --help`.
```

Comment:
Should auxiliary goals support `--opt` "option goals"? I lean towards no to keep the exposed API simpler...

Reply:
I had not intended on changing the `BuiltinGoal ` API. I pretty much did not have to disturb any of the existing option parsing logic for this PR thankfully. While the suggestion would simplify the public API, it would entail modifying the existing logic which has been undisturbed thus far and risks introducing bugs. Thoughts?